runengine-server.md

```diff
@@ -0,0 +1,96 @@
+configuration on where to save data e.t.c, and then ensuring that these are
+kept in-sync between GUIs / IPython profiles / Jupyter Notebooks, is painful. A
+benefit of going to a RunEngine server approach is you only need to launch the
+server with the desired configuration, then each of these processes only needs
+to know the information needed to know the location of the server.
+
+#### Experimental Control
+
+Centralizing the execution of scans can be beneficial to ensure that procedures
+are not adversely affecting each other. The fact that the server will not run
+plans in parallel would prevent operator errors where calibration routines are
+accidentally requested during data collection for example. This could be
+especially useful for automated tuning tasks that you want to run but do want
+to interfere during another user requested operation. 
+
+#### Experimental Gateway
+
+EPICS does not lend itself well to security, that is the nature of a
+distributed system. However, by setting up a RunEngine server that can be
+reached by a user, you can allow a specific restricted set of operations on a
+beamline without needing to give them permission to EPICS infrastructure at
+all. This could be incredibly useful for beamlines that are operated remotely
+via a web interface for example.
+
+## Issues and Pull Requests
+
+- https://github.com/bluesky/bluesky/pull/479
+
+## Affected Repositories
+
+- ophyd
+- bluesky
+
+## Open Design Decisions and Challenges
+
+#### Device Proxies
+Fortunately, `Msg` objects are trivial to serialize and ship over a wire, and
+for the most part the simple string and number arguments inside are as well.
+The one exception is how to correctly pass devices from client to server. The
+objects with live EPICS connections have to live on the server, the client has
+to be aware of which devices they can use in scans and how to specify each one.
+
+- The server could load a fixed set of devices. A user could then simply use
+  the "name" of devices instead of passing in the object itself. The server
+  would then be responsible for keeping a registry for lookup. This would require
+  names to be unique which is not currently enforced on any level.
+- The server could use [happi](https://github.com/pcdshub/happi) to load devices on request.
+- Some form of an `ophyd.DeviceProxy` could be created to exchange information
+  with the server version of the device. This would mean objects on the client
+  side could look very similar to the user as if they were actually making direct
+  control layer calls. There is enough information in `Component` that I believe
+  this is possible, but complex.
+
+#### Local Callbacks
+Currently, running callbacks in a separate process than the RunEngine is
+possible. I am of the belief that the API would need to be improved around this
+functionality in order to support this kind of running mode.
+
+#### Communication Layer
+What do we use to actually communicate with the server? ZMQ?
```

Comment:
I think it would not be hard to do two or more different protocols just to prove that they are easy to drop in.